### PR TITLE
docs: clarify common pitfall related to Vite's public folder (#100)

### PR DESCRIPTION
**Pull request by AleksandrHovhannisyan**
Let me know if any bits are unclear or poorly worded.

Changed it to `Common pitfalls` for consistency with 11ty's docs. ~Also, is it okay that the sidebar title is so long? Any shorter wording come to mind? Edit: `Common pitfalls`?~

### CODE REVIEW

File: docs/src/docs/common-pitfalls.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+1. Processing your site with 11ty and dumping the result into a temporary build folder (e.g., `.11ty-build-DcNVBN`).
+2. Processing the temporary build folder with Vite to bundle assets on the pages that reference them.
+3. Writing the temporary build folder to your 11ty output directory (e.g., `_site`)
+4. Deleting the temporary build folder.
+
+Vite is very strict about where it wants you to place your static assets—in a [dedicated `public` folder](https://vitejs.dev/guide/assets.html#the-public-directory). This allows it to distinguish static assets from ones that are referenced by your pages (like stylesheets) and to bundle the referenced assets while leaving the static ones alone. Any static asset that isn't referenced by a page and isn't in the `public` directory will be removed from the final build output. For this reason, you may find that certain files disappear from production builds, even though they were technically processed correctly by 11ty.
+
+Below are some scenarios where you may encounter this pitfall:
+
+- Self-hosted font files
```
Comment (Holben888) [inline docs/src/docs/common-pitfalls.md:22]: This is a _bit_ misleading I'm afraid. For instance, fonts actually don't need to go in the `public` folder, so long as they're referenced from some other file using a relative path or [import alias](https://deploy-preview-114--slinkity.netlify.app/docs/import-aliases/). For instance, If I have a CSS file that applies the `@font-face` like so:

```css
@font-face {
  font-family: Atkinson;
  src: url('./fonts/Atkinson/Atkinson-Hyperlegible-Regular.woff2'),
      url('./fonts/Atkinson/Atkinson-Hyperlegible-Regular.woff'),
      url('./fonts/Atkinson/Atkinson-Hyperlegible-Regular.ttf');
  font-display: swap;
}
```

...they'll get copied to the build output no problem! The problem only arrises for resources that _no one references via a relative path or import alias._

That admittedly wasn't clear to me after reading this section. I'd honestly suggest adding the example I listed above to explain what _does_ work.

File: docs/src/docs/common-pitfalls.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+Let's look at an example.
+
+### The old approach (11ty)
+
+Suppose you're using 11ty to auto-generate a `sitemap.xml` for your site. In 11ty, you'd create this file using your chosen templating language extension and set a permalink in its front matter relative to the output directory, like so:
+
+{% raw %}
+```liquid
+{% comment %}src/sitemap.liquid{% endcomment %}
```
Comment (Holben888) [inline docs/src/docs/common-pitfalls.md:34]:
```suggestion
{% comment %}src/sitemap.liquid{% endcomment %}
<!--src/sitemap.liquid-->
```

I don't think we're using the `comment` shortcode in our project. Is this a liquid feature? We're admittedly using nunjucks as our default language for `.md` files, so this shortcode doesn't get resolved.

File: docs/src/docs/common-pitfalls.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+permalink: /sitemap.xml
+---
+```
+{% endraw %}
+
+You would do this:
+
+{% raw %}
+```liquid
+{% comment %}src/sitemap.liquid{% endcomment %}
```
Comment (Holben888) [inline docs/src/docs/common-pitfalls.md:64]:
```suggestion
<!--src/sitemap.liquid-->
```

Ditto

File: docs/src/docs/common-pitfalls.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+---
+permalink: /public/sitemap.xml
+---
+```
+{% endraw %}
+
+Now, this happens:
+
+1. 11ty processes the template and writes it to a `public` folder nested inside the temporary build folder: `.11ty-build-DcNVBN/public/sitemap.xml`.
+2. Vite sees the `public` folder and copies it into your final build output directory, giving you `_site/public/sitemap.xml`.
```
Comment (Holben888) [inline docs/src/docs/common-pitfalls.md:74]:
```suggestion
2. Vite sees the `public` folder and copies it into your final build output directory, giving you `_site/sitemap.xml`. Note the nested `/public` disappears in the final build output!
```

☝️ Fun fact: `public` is a disappearing directory!

File: docs/src/docs/common-pitfalls.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+Similarly, if you're using font files on your site, you'll need to store them in a `public` folder outside your input directory and tell Eleventy to passthrough-copy them. This ensures that they make it into the temporary build folder, get picked up by Vite, and get written to the final output directory.
```
Comment (Holben888) [inline docs/src/docs/common-pitfalls.md:76]: Not sure about this section for the reasons I listed earlier on fonts... I think `robots.txt` and open graph images are better examples to reiterate here

File: docs/src/docs/common-pitfalls.md (hunk 1 above)
Comment (Holben888) [reply docs/src/docs/common-pitfalls.md:22]: Also worth stating:
1. You can configure `/public` to be any directory you like using a Vite Config (see [config options](https://deploy-preview-114--slinkity.netlify.app/docs/config/#vite's-vite.config.js))
2. You'll need to [passthrough copy](https://www.11ty.dev/docs/copy/) your public directory for it to exist in the final build output

**Review by Holben888 — COMMENTED** [review]
Have some comments on this! Just want to be thorough since it is a pretty confusing problem

File: docs/src/docs/common-pitfalls.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+## Certain files disappeared from my production build output folder
+
+In an 11ty site, you are free for the most part to structure your project however you like. You can put your static assets in a `public` folder outside your [11ty input directory]((https://www.11ty.dev/docs/config/#input-directory)) and pass-through copy them, but you don't need to—you could also store all of your assets inside your 11ty input directory. It doesn't matter which approach you take, so long as you configure your 11ty project accordingly and write your files to the correct locations. However, you need to take a slightly different approach in Slinkity due to the standards enforced by Vite.
+
+First, we need to understand that a Slinkity build consists of multiple steps:
+
+1. Processing your site with 11ty and dumping the result into a temporary build folder (e.g., `.11ty-build-DcNVBN`).
+2. Processing the temporary build folder with Vite to bundle assets on the pages that reference them.
+3. Writing the temporary build folder to your 11ty output directory (e.g., `_site`)
+4. Deleting the temporary build folder.
```
[the remark this answers is not in the document]
Comment (AleksandrHovhannisyan) [reply docs/src/docs/common-pitfalls.md:16]: That was mainly needed because of the phrasing before the list: `First, we need to understand that a Slinkity build consists of multiple steps:` (i.e., "A slinkity build consists of verbing this, verbing that, etc."). We can change it to be:

```
First, we need to understand how Slinkity builds work. In a typical build, we:
```

And then this suggestion works 👍

File: docs/src/docs/common-pitfalls.md (hunk 2 above)
Comment (AleksandrHovhannisyan) [reply docs/src/docs/common-pitfalls.md:34]: Yeah, this is a Liquid feature. I wanted to clarify the file's name using comments, but Prism didn't seem to like either `njk` or `nunjucks` as the language name, so I just went with Liquid.

File: docs/src/docs/common-pitfalls.md [hunk 7]
```diff
@@ -0,0 +1,76 @@
+---
+title: Common pitfalls
+---
+
+For the most part, you should be able to seamlessly migrate your site from 11ty to Slinkity with minimal reconfiguration. But there are a few changes that may catch you off guard. These gotchas are summarized below.
+
+## Certain files disappeared from my production build output folder
+
+In an 11ty site, you are free for the most part to structure your project however you like. You can put your static assets in a `public` folder outside your [11ty input directory]((https://www.11ty.dev/docs/config/#input-directory)) and pass-through copy them, but you don't need to—you could also store all of your assets inside your 11ty input directory. It doesn't matter which approach you take, so long as you configure your 11ty project accordingly and write your files to the correct locations. However, you need to take a slightly different approach in Slinkity due to the standards enforced by Vite.
```
[the remark this answers is not in the document]
Comment (AleksandrHovhannisyan) [reply docs/src/docs/common-pitfalls.md:9]: Looks good to me! Love the humor 😄 One nitpick: `there are a couple "gotchas"`.

(I really wish GitHub would let you suggest changes on suggested changes 😒)

File: docs/src/docs/common-pitfalls.md (hunk 4 above)
Comment (AleksandrHovhannisyan) [reply docs/src/docs/common-pitfalls.md:75]: Oh, I still see it in my build output folder. Is that a bug?

image

It still works because browsers just ignore the public bit, so a request to `/sitemap.xml` goes through successfully (in fact, attempting to access `/public/sitemap.xml` directly gives a 404.)

File: docs/src/docs/common-pitfalls.md (hunk 1 above)
Comment (AleksandrHovhannisyan) [reply docs/src/docs/common-pitfalls.md:23]: What are your thoughts on us splitting the explanations for the `public` folder out into a separate page? Too heavy handed?

File: docs/src/docs/common-pitfalls.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (Holben888) [reply docs/src/docs/common-pitfalls.md:9]: Haha yeah totally agree! The diffs could get crazy but an "edit before accepting suggestion" would be nice

File: docs/src/docs/common-pitfalls.md (hunk 1 above)
Comment (Holben888) [reply docs/src/docs/common-pitfalls.md]: Decision: split out to new "how we build your site" page. Then, add further sections on managing font faces as an example

**Comment by AleksandrHovhannisyan** [comment]
@Holben888 I went ahead and fleshed out the common pitfalls like we discussed. Feel free to suggest any edits!

Couple things I wanted to call out in particular:

## Plaintext fenced code blocks 

In some places, I wanted to show a sample tree view for a project. Fenced code blocks with no language looked a bit ugly, so I opted instead for the approach in 9856d08. It's not ideal, but it works:

image

I do something similar on my site. Let me know how you feel about this!

## Preload tags

I didn't mention how to handle these since I worried that it might get a little confusing. On the other hand, if users take the first approach (relying on Vite to auto-copy resources), they won't be able to use preload links since you need to know the hash ahead of time, but only Vite knows that. Unfortunately, from a local sandbox I set up, it seems like both import aliases and relative paths don't work for preloads `href`s and those strings end up getting written as-is.

Thoughts? Should we mention this?

File: docs/src/docs/common-pitfalls.md [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+```plaintext
+├── fonts
+│   └── Atkinson-Hyperlegible-Regular.woff2
+├── src
+├── _includes
+│   │   └── layout.njk
+│   ├── index.md
+├── styles
+│   ├── index.css
+```
```
Comment (AleksandrHovhannisyan) [inline docs/src/docs/common-pitfalls.md:79]: On my site, so that I don't have to remember to add `plaintext` for plaintext code blocks, I opted to remove 11ty's syntax highlighter plugin and replace it with markdown-it-prism, setting the default language to `plaintext`:

https://github.com/AleksandrHovhannisyan/aleksandrhovhannisyan.com/blob/master/config/plugins/markdown.js#L15

Thoughts on doing this?

File: docs/styles/prism-synthwave-84.css [hunk 9]
```diff
@@ -36,6 +36,11 @@ pre[class*='language-'] {
   overflow: auto;
 }
 
+/* Plaintext code blocks */
+:is(code,pre)[class='language-plaintext'] {
+  color: #fff5f6;
+}
+
```
Comment (AleksandrHovhannisyan) [inline docs/styles/prism-synthwave-84.css:43]: In some places, I wanted to show a sample tree view for a project. Fenced code blocks with no language looked a bit ugly, so I opted instead for the approach in 459f3d4. It's not ideal because we need to remember to add `plaintext` to code blocks, but it works:

image

On my site, so that I don't have to remember to add `plaintext` for plaintext code blocks, I opted to remove 11ty's syntax highlighter plugin and replace it with markdown-it-prism, setting the default language to `plaintext`:

https://github.com/AleksandrHovhannisyan/aleksandrhovhannisyan.com/blob/master/config/plugins/markdown.js#L15

Thoughts on doing this?

File: docs/src/docs/common-pitfalls.md [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+A typical 11ty site has many static resources, including fonts, images, and more. Some of these (like images) may need to be processed manually using 11ty, but others (like fonts) just need to be copied to the build output directory.
+
+In Slinkity, there are two approaches you can take to ensure that static resources correctly make their way to the final build output:
+
+1. Using relative paths or import aliases to the resources on a page.
+2. Passthrough-copying the resources to a `/public` folder.
+
+#### Approach 1: Path resolution
+
+In the first approach, we lean on Vite's ability to resolve dependencies and resources referenced by our templates. In this case, if you want your font files to end up in your site's final build output, all you need to do is reference those font files somewhere in your CSS. You can do this using either relative paths (not absolute!) or [import aliases](/docs/import-aliases).
```
Comment (AleksandrHovhannisyan) [inline docs/src/docs/common-pitfalls.md:66]: If users take this approach (relying on Vite to auto-copy resources), they won't be able to use preload links since you need to know the hash ahead of time, but only Vite knows that. Unfortunately, from a local sandbox I set up, it seems like both import aliases and relative paths don't work for preloads `href`s and those strings end up getting written as-is.

Thoughts? Should we mention this?

Comment (AleksandrHovhannisyan) [reply docs/src/docs/common-pitfalls.md:66]: Here's the sandbox branch/file in question if you want to take a look: https://github.com/AleksandrHovhannisyan/slinkity-sandbox/blob/import-aliases-for-static-resources/src/_includes/layout.njk#L9

I could also open a separate ticket for that if you want.

Comment (Holben888) [reply docs/src/docs/common-pitfalls.md:66]: And yes, I think we can add a `> quoteblock` to explain that caveat. We'll say that Vite can't resolve relative paths or import aliases from preload tags, so if you need them, we suggest approach 2

File: docs/styles/prism-synthwave-84.css (hunk 9 above)
Comment (AleksandrHovhannisyan) [reply docs/styles/prism-synthwave-84.css:43]: I use the `tree` command in Bash for this (`tree -I node_modules`).

Not sure if the extra characters add unnecessary noise for screen readers, though; I haven't tested it. I could use indentation here if you'd like—up to you!

Comment (AleksandrHovhannisyan) [reply docs/styles/prism-synthwave-84.css:43]: I suppose we could also just override the colors for `language-bash` and use that instead.

Comment (Holben888) [reply docs/styles/prism-synthwave-84.css:43]: Yes to `plaintext` as a "default." I mainly used `bash` to add comments in-line where necessary. But for file trees like this, I agree with you

Comment (AleksandrHovhannisyan) [reply docs/styles/prism-synthwave-84.css:43]: Sounds good, I'll do that as a separate PR to keep this one more manageable

Comment (Holben888) [reply docs/styles/prism-synthwave-84.css]: Awesome! I'll also get this in once I add an initial draft for "how we build your site." Prioritizing work on renderers right now but it's on my todos for the extended weekend ❤️

File: www/src/docs/common-pitfalls.md [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+@font-face {
+  font-family: Atkinson;
+  src: url(/assets/Atkinson-Hyperlegible-Regular.efe52c57.woff2);
+  font-display: swap;
+}
+```
+
+> ⚠️ One downside to this approach: Vite can't resolve relative paths or import aliases from `preload` tags. If you need these on your site, we recommend going with the passthrough-copy approach instead.
+
+#### Approach 2: Passthrough-copying a `public` folder
```
Comment (Holben888) [inline www/src/docs/common-pitfalls.md:157]: On closer inspection, I realized... this section is wrong! There's no need to passthrough copy to a `public` folder specifically, as long as your absolute paths reference whatever you're passthrough copying. 

Take these fonts for instance. Your `@font-face src` can happily point to `/fonts/Atkinson-Hyperlegible-Regular.woff2` as long as your passthrough copying to `fonts/`. Vite will discover that your CSS is referencing the file and copy this to the output. Right now, `robots.txt` and `sitemap.xml` are the only examples I can think of that would require a passthrough copy on `public/`. I think I'll rework this section to accomodate (but preserve your lovely folder trees 🙃

Comment (AleksandrHovhannisyan) [reply www/src/docs/common-pitfalls.md]: Huh, you're right! Just tested in a sandbox. Apparently Vite _is_ clever enough to pick up on that relationship. It even updates the font-face CSS paths to correctly reference `/assets/whatever.woff2`.

**Comment by AleksandrHovhannisyan** [comment]
@Holben888 Lgtm! For the last example, I think we want to un-nest `robot.txt` from `assets`.

**Review by Holben888 — APPROVED** [review]
At last... we crushed this. Ship it! 🚢
